dataikuapi/dss/ml.py

```diff
@@ -923,6 +923,26 @@ def __init__(self, raw_settings, hyperparameter_search_params):
         self._hyperparameter_search_params = hyperparameter_search_params
         self._hyperparameters_registry = dict()
 
+    def __setattr__(self, key, value):
+        if key in {"_hyperparameters_registry", "_hyperparameter_search_params"} or key in {"enabled", "strategy"}:
+            # attributes and properties of the PredictionAlgorithmSettings object must be handled separately
+            super(PredictionAlgorithmSettings, self).__setattr__(key, value)
+        elif key in self._hyperparameters_registry or (key=="lambda_" and "lambda" in self._hyperparameters_registry):
+            if isinstance(value, HyperparameterSettings):
+                # call from a PredictionAlgorithmSettings child's __init__
+                super(PredictionAlgorithmSettings, self).__setattr__(key, value)
+            else:
+                # syntactic sugars
+                target = self._hyperparameters_registry[key]
+                if isinstance(target, (SingleValueHyperparameterSettings, SingleCategoryHyperparameterSettings)):
+                    target.set_value(value)
+                elif isinstance(target, CategoricalHyperparameterSettings):
+                    target.set_values(value)
+                else:
+                    raise Exception("Invalid type for assignment of a NumericalHyperparameterSettings object")
+        else:
+            raise Exception("Unknown hyperparameter: \"{}\"".format(key))
+
```

Comment:
I don't think this works for the `__init__`, because the keys are added to the registry in the `__init__`, so they won't be there at first execution, and you will directly go to the `else` clause.

Can't we simply check if `hasattr(self, key)` to distinguish between the `__init__` and non `__init__` case. Then:
* If `not hasattr(self, key)` =>  We are in `__init__` (or `key` is a new attribute that we don't care about) => `super(PredictionAlgorithmSettings, self).__setattr__(key, value)`
* Otherwise => we are not in `__init__` for sure => check if key is in registry
    * if not in registry so this is not a hyperparam, do the standard thing => `super(PredictionAlgorithmSettings, self).__setattr__(key, value)` 
    * If in registry, this is a hyperparam=> check `target = self._hyperparameters_registry[key]`
        * If `target` is an instance of one of the three specific classes, use set value(s) method (and also would be nice to add type check in the `set_values` and `set_value` methods (it is already the case in `SingleValueHyperparameterSettings.set_value`) )
        * Otherwise raise an Exception, because we don't want the API users to manually modify the hyperparams
        
What do you think ?

Reply:
Actually it does work because the call to `self._register_*_hyperparameter` (and thus the registration to `_hyperparameters_registry`) is executed _before_ the call to `__setattr__` in the children's `__init__`...
But the flow you're suggesting might be less ugly / more understandable and maintainable, so let me try and test it.

Reply:
OK much better indeed IMO, thanks for the suggestion !